docs/content/sips/011-component-versioning.md

```diff
@@ -0,0 +1,41 @@
+---
+
+Summary: Embed version information into Spin components for improved compatibility diagnostics.
+
+Owner(s): joel.dice@fermyon.com
+
+Created: Jan 24, 2023
+
+## Background
+
+Currently, Spin has no way to determine which version of SDK the components of an application were built with, which leads to obscure, hard-to-understand error messages for ABI and/or API incompatibilities.  For example, a component built with a Spin 0.4 SDK won't run on Spin 0.7 due to `wit-bindgen` ABI changes, leading to a variety of confusing runtime errors like out-of-bounds memory traps, assertion errors about invalid values, etc.  In addition, new APIs were added between 0.4 and 0.7, and thus a 0.7 component might not run on 0.4 even if the ABI had not changed.  These errors are usually a bit easier to understand (e.g. missing imports), but still don't point clearly to a resolution.  Ideally, the message presented to the user would be something like, "This component was built with a Spin 0.7 SDK and cannot be run using this version of Spin (0.4) -- please use Spin 0.7 or later."
+
+Note that when we say "SDK" in this document, we mean either one of the official Spin SDKs _or_ raw bindings generated from the `WIT` files and the `wit-bindgen` version shipped with a given version of Spin, e.g. for languages which do not yet have an "official" SDK.
+
+## Proposal
+
+Although `WIT` does not yet have a way to express API versions, we can emulate it by introducing a new interface called `spin-sdk-version.wit` containing a single function to be exported by the guest component:
+
+```
+record version-details {
+  commmit-hash: string,
+  sdk-language: string
+}
+
+spin-sdk-version-$MAJOR-$MINOR: func() -> version-details
+```
+
+where `$MAJOR` and `$MINOR` are the Spin major and minor version numbers the SDK targets. The intention behind embedding these numbers in the name of the function is that it can be checked statically without instantiating or running the component, and it is immune to ABI changes.  The `version-details` record provides additional details about the SDK used to build the component.  Getting that information requires actually running the component, which may not be possible if the host does not support the ABI for that version, so it won't be useful in all scenarios.  (TODO: should we just encode everything in the name?)
+
+When Spin deploys or runs an application, it will have an opportunity to check the versions of its component(s) for compatibility, resulting in one of the following scenarios:
+
+- *No version found*: accept and run the component, raising an error if an incompatibility is detected.  This matches the current behavior.
+
+- *Version found, and it's within the range that this build of Spin supports*: accept and run the component.  Presumably no ABI or API incompatibilities are possible in this scenario, unless the component is reporting an inaccurate version or a breaking SDK change was made without bumping the version.
```

Comment:
"Within the range that this build of Spin supports" sounds like a possibly elusive target.  Consider the case that Spin 0.11 scans the exports and find `spin-sdk-version-0.9`.  How does it know whether it supports 0.9?  Does it have a table?  How is that table maintained?

This gets simpler once we get into non-zero major versions because within a major version we know that if Spin's minor version is higher than the exported version then we're safe.  We would still need to maintain a table of compatible major versions but hopefully that would be less onerous given the release cadence.

Then again, at non-zero major versions I really hope we are introducing some sort of rolling upgrade policy for APIs, meaning this would be needed only when the ABI changed, and when deprecated APIs were finally put out to pasture, all of which would hopefully be infrequent and well trailed.

Reply:
I've been deliberately vague about how a given Spin version decides which SDK versions are supported, given that the primary goals of this SIP are:

1. Start recording versions in components so we have _something_ to work with.
2. Provide better error messages when imports are missing and/or the ABI is broken.

I'll add a "future discussion" section to discuss various ways future Spin versions could make compatibility decisions vs. the "exact match only" approach I imagine we would start with.

Reply:
I've added a `Future work` section to address these topics.  Let me know if it's still not clear.